File: client/mia_android/app/src/main/kotlin/com/mia/companion/ui/screens/AuthScreen.kt
package com.mia.companion.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material.icons.outlined.Visibility
import androidx.compose.material.icons.outlined.VisibilityOff
import androidx.compose.material.icons.outlined.WifiOff
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import com.mia.companion.data.ApiClient
import com.mia.companion.ui.theme.MiaColors
import com.mia.companion.ui.theme.MiaTypography
import kotlinx.coroutines.launch

@Composable
fun AuthScreen(onLoggedIn: () -> Unit) {
    var isRegister by remember { mutableStateOf(false) }
    var username by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var serverReachable by remember { mutableStateOf<Boolean?>(null) }
    var showPassword by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    LaunchedEffect(Unit) {
        serverReachable = ApiClient.checkHealth()
    }

    val fieldColors = OutlinedTextFieldDefaults.colors(
        focusedBorderColor = MiaColors.Accent,
        unfocusedBorderColor = MiaColors.MiaBubble,
        focusedContainerColor = MiaColors.Surface,
        unfocusedContainerColor = MiaColors.Surface,
    )

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(MiaColors.Background),
    ) {
        if (serverReachable == false) {
            OfflineBanner(onRetry = { scope.launch { serverReachable = ApiClient.checkHealth() } })
        }
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 28.dp, vertical = 28.dp),
        ) {
            Spacer(Modifier.height(48.dp))
            Text("mia", style = MiaTypography.serifTitle(44f))
            Spacer(Modifier.height(8.dp))
            Text(
                if (isRegister) "create your account" else "welcome back",
                style = MiaTypography.inter(15f, FontWeight.Medium, MiaColors.StatusPink),
            )
            Spacer(Modifier.height(40.dp))

            OutlinedTextField(
                value = username,
                onValueChange = { username = it },
                placeholder = { Text("username") },
                leadingIcon = { Icon(Icons.Outlined.Person, null, Modifier.size(22.dp)) },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(28.dp),
                colors = fieldColors,
            )
            Spacer(Modifier.height(14.dp))
            OutlinedTextField(
                value = password,
                onValueChange = { password = it },
                placeholder = { Text("password") },
                leadingIcon = { Icon(Icons.Outlined.Lock, null, Modifier.size(22.dp)) },
                trailingIcon = {
                    IconButton(onClick = { showPassword = !showPassword }) {
                        Icon(
                            if (showPassword) Icons.Outlined.VisibilityOff else Icons.Outlined.Visibility,
                            null,
                            Modifier.size(22.dp),
                        )
                    }
                },
                visualTransformation = if (showPassword) VisualTransformation.None else PasswordVisualTransformation(),
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(28.dp),
                colors = fieldColors,
            )

            error?.let { msg ->
                Spacer(Modifier.height(16.dp))
                Text(
                    msg,
                    style = MiaTypography.inter(13f, color = MiaColors.AccentDeep, lineHeight = 17.55f),
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(MiaColors.AccentDeep.copy(alpha = 0.08f), RoundedCornerShape(16.dp))
                        .border(1.dp, MiaColors.AccentDeep.copy(alpha = 0.2f), RoundedCornerShape(16.dp))
                        .padding(14.dp),
                )
            }

            Spacer(Modifier.height(28.dp))
            Button(
                onClick = {
                    val u = username.trim()
                    if (u.length < 3 || !Regex("^[a-zA-Z0-9_]+$").matches(u)) {
                        error = "at least 3 characters, letters/numbers/underscore only"
                        return@Button
                    }
                    if (password.length < 6) {
                        error = "at least 6 characters"
                        return@Button
                    }
                    scope.launch {
                        loading = true
                        error = null
                        try {
                            if (isRegister) ApiClient.register(u, password)
                            else ApiClient.login(u, password)
                            onLoggedIn()
                        } catch (e: Exception) {
                            error = friendlyAuthError(e)
                        } finally {
                            loading = false
                        }
                    }
                },
                enabled = !loading && serverReachable != false,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(52.dp),
                shape = RoundedCornerShape(28.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = MiaColors.Accent,
                    disabledContainerColor = MiaColors.Accent.copy(alpha = 0.45f),
                ),
            ) {
                if (loading) {
                    CircularProgressIndicator(
                        modifier = Modifier.size(22.dp),
                        color = MiaColors.Surface,
                        strokeWidth = 2.dp,
                    )
                } else {
                    Text(if (isRegister) "sign up" else "log in")
                }
            }
            TextButton(
                onClick = {
                    isRegister = !isRegister
                    error = null
                },
                enabled = !loading,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Text(
                    if (isRegister) "already have an account? log in" else "new here? create account",
                    style = MiaTypography.inter(15f, FontWeight.Medium, MiaColors.AccentDeep),
                )
            }
        }
    }
}

@Composable
private fun OfflineBanner(onRetry: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(MiaColors.AccentDeep.copy(alpha = 0.12f))
            .padding(horizontal = 16.dp, vertical = 10.dp),
    ) {
        Icon(Icons.Outlined.WifiOff, null, tint = MiaColors.AccentDeep, modifier = Modifier.size(18.dp))
        Spacer(Modifier.width(10.dp))
        Text(
            "can't reach server at ${ApiClient.apiBaseUrl}",
            style = MiaTypography.inter(12f, color = MiaColors.AccentDeep, lineHeight = 15.6f),
            modifier = Modifier.weight(1f),
        )
        TextButton(onClick = onRetry) {
            Text("retry", style = MiaTypography.inter(14f, FontWeight.SemiBold, MiaColors.AccentDeep))
        }
    }
}

private fun friendlyAuthError(e: Exception): String {
    val raw = e.message?.replace("Exception: ", "") ?: "something went wrong"
    if (raw.contains("Cannot reach server", ignoreCase = true)) {
        return "can't reach mia's server. on your phone open ${ApiClient.apiBaseUrl}/health — if that fails, set private DNS to automatic or dns.google, then try again."
    }
    if (raw.contains("Username already taken", ignoreCase = true)) return "that username is taken — try another."
    if (raw.contains("Invalid username or password", ignoreCase = true)) return "wrong username or password."
    return raw
}

File: client/mia_companion/lib/screens/auth_screen.dart
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';

import '../config.dart';
import '../data/mia_profile.dart';
import '../services/api_service.dart';
import '../theme/mia_theme.dart';
import 'chat_screen.dart';

class AuthScreen extends StatefulWidget {
  const AuthScreen({super.key, this.initialRegister = true});

  final bool initialRegister;

  @override
  State<AuthScreen> createState() => _AuthScreenState();
}

class _AuthScreenState extends State<AuthScreen> {
  late bool _isRegister = widget.initialRegister;
  final _username = TextEditingController();
  final _password = TextEditingController();
  final _formKey = GlobalKey<FormState>();
  bool _loading = false;
  String? _error;
  bool? _serverReachable;
  bool _obscurePassword = true;

  @override
  void initState() {
    super.initState();
    _checkServer();
  }

  Future<void> _checkServer() async {
    final ok = await ApiService.instance.checkHealth();
    if (mounted) setState(() => _serverReachable = ok);
  }

  @override
  void dispose() {
    _username.dispose();
    _password.dispose();
    super.dispose();
  }

  Future<void> _submit() async {
    if (!_formKey.currentState!.validate()) return;

    FocusScope.of(context).unfocus();
    setState(() {
      _loading = true;
      _error = null;
    });

    try {
      if (_isRegister) {
        await ApiService.instance.register(
          _username.text.trim(),
          _password.text,
        );
      } else {
        await ApiService.instance.login(
          _username.text.trim(),
          _password.text,
        );
      }

      if (!mounted) return;
      Navigator.of(context).pushReplacement(
        MaterialPageRoute(builder: (_) => const ChatScreen()),
      );
    } catch (e) {
      setState(() => _error = _friendlyError(e));
    } finally {
      if (mounted) setState(() => _loading = false);
    }
  }

  String _friendlyError(Object e) {
    final raw = e.toString().replaceFirst('Exception: ', '');
    if (raw.contains('Cannot reach server')) {
      if (isProductionApi) {
        return 'can\'t reach ${MiaProfile.name.toLowerCase()}\'s server. on your phone open $resolvedApiBaseUrl/health — if that fails, set private DNS to automatic or dns.google, then try again.';
      }
      return 'can\'t reach the server. run npm run dev on your mac, then use flutter run with your mac\'s ip.';
    }
    if (raw.contains('Username already taken')) {
      return 'that username is taken — try another.';
    }
    if (raw.contains('Invalid username or password')) {
      return 'wrong username or password.';
    }
    return raw;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      resizeToAvoidBottomInset: true,
      body: Stack(
        fit: StackFit.expand,
        children: [
          Image.asset(
            'assets/images/auth_background.png',
            fit: BoxFit.cover,
            alignment: Alignment.centerLeft,
          ),
          DecoratedBox(
            decoration: BoxDecoration(
              gradient: LinearGradient(
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
                colors: [
                  MiaColors.background.withValues(alpha: 0.15),
                  MiaColors.background.withValues(alpha: 0.45),
                  MiaColors.background.withValues(alpha: 0.88),
                ],
                stops: const [0.0, 0.42, 1.0],
              ),
            ),
          ),
          DecoratedBox(
            decoration: BoxDecoration(
              gradient: LinearGradient(
                begin: Alignment.centerLeft,
                end: Alignment.centerRight,
                colors: [
                  Colors.transparent,
                  MiaColors.background.withValues(alpha: 0.35),
                  MiaColors.background.withValues(alpha: 0.82),
                ],
                stops: const [0.0, 0.55, 1.0],
              ),
            ),
          ),
          SafeArea(
            child: Column(
              children: [
                if (_serverReachable == false)
                  _OfflineBanner(onRetry: _checkServer),
                Expanded(
                  child: LayoutBuilder(
                    builder: (context, constraints) {
                      final wide = constraints.maxWidth >= 720;
                      return Align(
                        alignment:
                            wide ? Alignment.centerRight : Alignment.bottomCenter,
                        child: SingleChildScrollView(
                          padding: EdgeInsets.fromLTRB(
                            wide ? constraints.maxWidth * 0.42 : 28,
                            wide ? 24 : 16,
                            28,
                            28,
                          ),
                          child: ConstrainedBox(
                            constraints: BoxConstraints(
                              maxWidth: wide ? 420 : double.infinity,
                            ),
                            child: DecoratedBox(
                              decoration: BoxDecoration(
                                color: MiaColors.surface.withValues(alpha: 0.72),
                                borderRadius: BorderRadius.circular(28),
                                border: Border.all(
                                  color: MiaColors.miaBubble.withValues(alpha: 0.6),
                                ),
                                boxShadow: [
                                  BoxShadow(
                                    color: MiaColors.accentDeep.withValues(alpha: 0.08),
                                    blurRadius: 24,
                                    offset: const Offset(0, 8),
                                  ),
                                ],
                              ),
                              child: Padding(
                                padding: const EdgeInsets.fromLTRB(24, 28, 24, 24),
                                child: _buildForm(),
                              ),
                            ),
                          ),
                        ),
                      );
                    },
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildForm() {
    return Form(
      key: _formKey,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Text('zara', style: MiaTheme.serifTitle(size: 44)),
          const SizedBox(height: 8),
          Text(
            _isRegister ? 'create your account' : 'welcome back',
            style: GoogleFonts.inter(
              color: MiaColors.statusPink,
              fontSize: 15,
              fontWeight: FontWeight.w500,
            ),
          ),
          const SizedBox(height: 32),
          TextFormField(
            controller: _username,
            decoration: const InputDecoration(
              hintText: 'username',
              prefixIcon: Icon(Icons.person_outline, size: 22),
            ),
            textInputAction: TextInputAction.next,
            autocorrect: false,
            validator: (v) {
              final t = v?.trim() ?? '';
              if (t.length < 3) return 'at least 3 characters';
              if (!RegExp(r'^[a-zA-Z0-9_]+$').hasMatch(t)) {
                return 'letters, numbers, underscore only';
              }
              return null;
            },
          ),
          const SizedBox(height: 14),
          TextFormField(
            controller: _password,
            obscureText: _obscurePassword,
            decoration: InputDecoration(
              hintText: 'password',
              prefixIcon: const Icon(Icons.lock_outline, size: 22),
              suffixIcon: IconButton(
                icon: Icon(
                  _obscurePassword
                      ? Icons.visibility_outlined
                      : Icons.visibility_off_outlined,
                  size: 22,
                ),
                onPressed: () => setState(
                  () => _obscurePassword = !_obscurePassword,
                ),
              ),
            ),
            onFieldSubmitted: (_) => _submit(),
            validator: (v) {
              if ((v?.length ?? 0) < 6) {
                return 'at least 6 characters';
              }
              return null;
            },
          ),
          if (_error != null) ...[
            const SizedBox(height: 16),
            Container(
              padding: const EdgeInsets.all(14),
              decoration: BoxDecoration(
                color: MiaColors.accentDeep.withValues(alpha: 0.08),
                borderRadius: BorderRadius.circular(16),
                border: Border.all(
                  color: MiaColors.accentDeep.withValues(alpha: 0.2),
                ),
              ),
              child: Text(
                _error!,
                style: GoogleFonts.inter(
                  color: MiaColors.accentDeep,
                  fontSize: 13,
                  height: 1.35,
                ),
              ),
            ),
          ],
          const SizedBox(height: 28),
          FilledButton(
            onPressed: (_loading || _serverReachable == false) ? null : _submit,
            child: _loading
                ? const SizedBox(
                    height: 22,
                    width: 22,
                    child: CircularProgressIndicator(
                      strokeWidth: 2,
                      color: Colors.white,
                    ),
                  )
                : Text(_isRegister ? 'sign up' : 'log in'),
          ),
          const SizedBox(height: 16),
          TextButton(
            onPressed: _loading
                ? null
                : () => setState(() {
                      _isRegister = !_isRegister;
                      _error = null;
                    }),
            child: Text(
              _isRegister
                  ? 'already have an account? log in'
                  : 'new here? create account',
              style: GoogleFonts.inter(
                color: MiaColors.accentDeep,
                fontWeight: FontWeight.w500,
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _OfflineBanner extends StatelessWidget {
  const _OfflineBanner({required this.onRetry});

  final VoidCallback onRetry;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: MiaColors.accentDeep.withValues(alpha: 0.12),
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 10),
        child: Row(
          children: [
            const Icon(Icons.wifi_off_rounded, size: 18, color: MiaColors.accentDeep),
            const SizedBox(width: 10),
            Expanded(
              child: Text(
                'can\'t reach server at $resolvedApiBaseUrl',
                style: GoogleFonts.inter(
                  fontSize: 12,
                  color: MiaColors.accentDeep,
                  height: 1.3,
                ),
              ),
            ),
            TextButton(
              onPressed: onRetry,
              child: Text(
                'retry',
                style: GoogleFonts.inter(
                  color: MiaColors.accentDeep,
                  fontWeight: FontWeight.w600,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
